### PR TITLE
`Paywalls`: handle missing paywalls gracefully

### PR DESCRIPTION
**Pull request by NachoSoto**
### Changes:
- Improved `PaywallView` API: now there's only 2 constructors (with optional `Mode` parameters):
    - `PaywallView()`
    - `PaywallView(offering:)`
- New `PaywallData.default` as a fallback when trying to present a paywall with missing data (either because it failed to decode, or it's missing)
- Extracted error state handling to `ErrorDisplay` (used by `PaywallView` and `AsyncButton` now). It can optionally dismiss the presenting sheet.
- Handling offering loading errors in `PaywallView`
- Improved `DebugErrorView` to allow displaying a fallback view:
```swift
DebugErrorView(
    "Offering '\(offering.identifier)' has no configured paywall.\n" +
    "The displayed paywall contains default configuration.\n" +
    "This error will be hidden in production.",
    releaseBehavior: .replacement(
        AnyView(
            LoadedOfferingPaywallView(
                offering: offering,
                paywall: .default,
                mode: mode,
                introEligibility: checker,
                purchaseHandler: purchaseHandler
            )
        )
    )
)
```
- Added `LoadingPaywallView` as a placeholder view during loading
- Improved `MultiPackageTemplate` and `RemoteImage` to fix some layout issues during transitions
- Added transition animations between loading state and loaded paywall

### CODE REVIEW

File: RevenueCatUI/Views/DebugErrorView.swift [hunk 1]
```diff
@@ -19,6 +19,7 @@ struct DebugErrorView: View {
 
         case emptyView
         case fatalError
+        case replacement(AnyView)
```
Comment (NachoSoto) [inline RevenueCatUI/Views/DebugErrorView.swift:22]: 💥

File: RevenueCatUI/Views/AsyncButton.swift [hunk 2]
```diff
@@ -36,60 +36,14 @@ struct AsyncButton<Label>: View where Label: View {
...
-        if let errorCode = self.underlyingError as? ErrorCode {
-            return errorCode.description
-        } else {
-            return self.underlyingError.localizedDescription
-        }
-    }
-
-    var recoverySuggestion: String? {
-        self.underlyingError.localizedRecoverySuggestion
+        .displayError(self.$error)
```
Comment (NachoSoto) [inline RevenueCatUI/Views/AsyncButton.swift:46]: Same thing in just one line.

File: RevenueCatUI/Views/LoadingPaywallView.swift [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+@available(iOS 16.0, macOS 13.0, tvOS 16.0, *)
+struct LoadingPaywallView: View {
+
+    var body: some View {
+        LoadedOfferingPaywallView(
+            offering: .init(
+                identifier: Self.offeringIdentifier,
+                serverDescription: "",
+                metadata: [:],
+                paywall: .default,
```
Comment (NachoSoto) [inline RevenueCatUI/Views/LoadingPaywallView.swift:21]: The loading view reuses the same `.default` paywall.

File: RevenueCatUI/Views/LoadingPaywallView.swift [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+                    Self.annualPackage
+                ]
+            ),
+            paywall: .default,
+            mode: .fullScreen,
+            introEligibility: Self.introEligibility,
+            purchaseHandler: Self.purchaseHandler
+        )
+        .disabled(true)
+        .redacted(reason: .placeholder)
```
Comment (NachoSoto) [inline RevenueCatUI/Views/LoadingPaywallView.swift:34]: But "redacted":
Simulator Screenshot - iPhone 14 Pro Max - 2023-07-21 at 09 34 54

File: RevenueCatUI/Views/LoadingPaywallView.swift [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+            uniqueKeysWithValues: packages.map { ($0, .unknown) }
+        )
+    })
+    static let purchaseHandler: PurchaseHandler = .init(
+        purchase: { _ in
+            fatalError("Should not be able to purchase")
+        },
+        restorePurchases: {
+            fatalError("Should not be able to purchase")
+        }
```
Comment (NachoSoto) [inline RevenueCatUI/Views/LoadingPaywallView.swift:53]: Important to create these so we don't attempt to purchase fake packages.
The view is marked as `.disabled()` to make this impossible.

File: RevenueCatUI/Views/RemoteImage.swift [hunk 6]
```diff
@@ -37,10 +40,22 @@ struct RemoteImage: View {
...
-                    .hidden()
+                Group {
+                    if let aspectRatio {
+                        self.placeholderView
+                            .aspectRatio(aspectRatio, contentMode: .fit)
+                    } else {
+                        self.placeholderView
+                    }
+                }
+                .frame(maxWidth: self.maxWidth)
```
Comment (NachoSoto) [inline RevenueCatUI/Views/RemoteImage.swift:51]: This ensures consistent layout when transitioning from loading to loaded.

File: Tests/TestingApps/SimpleApp/SimpleApp/SimpleApp.swift [hunk 7]
```diff
@@ -54,40 +50,6 @@ struct SimpleApp: App {
         }
     }
 
-    @ViewBuilder
-    private var paywallView: some View {
```
Comment (NachoSoto) [inline Tests/TestingApps/SimpleApp/SimpleApp/SimpleApp.swift:58]: All this boilerplate is gone and no longer necessary for our users either 🎉

File: RevenueCatUI/PaywallView.swift [hunk 8]
```diff
@@ -45,40 +46,43 @@ public struct PaywallView: View {
...
+                self.paywallView(for: offering,
+                                 checker: checker,
+                                 purchaseHandler: purchaseHandler)
+                .transition(Self.transition)
             } else {
-                self.loadingView
+                LoadingPaywallView()
+                .transition(Self.transition)
                     .task {
-                        // Fix-me: better error handling
```
Comment (NachoSoto) [inline RevenueCatUI/PaywallView.swift:79]: Fixed.

**Comment by NachoSoto** [comment]
TODO: expose `TestStoreProduct` on release builds to fix this.

File: RevenueCatUI/Resources/background.jpg
Comment (joshdholtz) [inline RevenueCatUI/Resources/background.jpg:1]: Do we have the rights to distribute this background? 😇

File: RevenueCatUI/Templates/MultiPackageTemplate.swift [hunk 9]
```diff
@@ -60,6 +60,7 @@ private struct MultiPackageTemplateContent: View {
             .frame(maxWidth: .infinity, maxHeight: .infinity)
             .background {
                 self.backgroundImage
+                    .unredacted()
```
Comment (joshdholtz) [inline RevenueCatUI/Templates/MultiPackageTemplate.swift:63]: What was this added for? 🤔

File: RevenueCatUI/Views/LoadingPaywallView.swift (hunk 4 above)
Comment (joshdholtz) [reply RevenueCatUI/Views/LoadingPaywallView.swift:34]: 😍

File: Tests/RevenueCatUITests/__Snapshots__/OtherPaywallViewTests/iOS16-testDefaultPaywall.1.png
Comment (joshdholtz) [inline Tests/RevenueCatUITests/__Snapshots__/OtherPaywallViewTests/iOS16-testDefaultPaywall.1.png:1]: Did we want this to be blurred? 🤷‍♂️

**Review by joshdholtz — APPROVED** [review]
Code looks good! A few small questions but nothing blocking the merge 😊

File: Tests/RevenueCatUITests/__Snapshots__/OtherPaywallViewTests/iOS16-testDefaultPaywall.1.png
Comment (NachoSoto) [reply Tests/RevenueCatUITests/__Snapshots__/OtherPaywallViewTests/iOS16-testDefaultPaywall.1.png:1]: It's not blurred, it's "redacted" (this answers your other question too:
Simulator Screenshot - iPhone 14 Pro Max - 2023-07-21 at 09 34 54

Happy to change this though to a blur 🤷🏻 we can polish this later.

File: RevenueCatUI/Templates/MultiPackageTemplate.swift (hunk 9 above)
Comment (NachoSoto) [reply RevenueCatUI/Templates/MultiPackageTemplate.swift:63]: See below.

File: RevenueCatUI/Resources/background.jpg
Comment (NachoSoto) [reply RevenueCatUI/Resources/background.jpg:1]: Hmm good point 🤦🏻 this is just WIP, I'm sure we'll do something else, but let me get a better temporary image for the blurred background for now.

**Pull request merged**